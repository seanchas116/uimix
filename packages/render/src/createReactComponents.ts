import { ProjectJSON } from "@uimix/node-data";
import React from "react";
import { Component, Node, Project } from "./Project.js";

export function createReactComponents(
  projectJSON: ProjectJSON,
  imageURLs: Record<string, string>,
  foreignModules: Record<string, Record<string, any>>
): Map<string, React.FC<ComponentProps>> {
  const project = new Project(projectJSON);
  const assets: Assets = {
    imageURLs: new Map(Object.entries(imageURLs)),
    foreignModules: new Map(Object.entries(foreignModules)),
  };

  const reactComponents = new Map<string, React.FC<ComponentProps>>();
  for (const component of project.components.values()) {
    reactComponents.set(
      component.name,
      createReactComponent(component, assets)
    );
  }

  return reactComponents;
}

interface Assets {
  imageURLs: Map<string, string>;
  foreignModules: Map<string, Record<string, any>>;
}

type ComponentProps = JSX.IntrinsicElements["div"] & {
  overrides?: { [key: string]: ComponentProps };
};

class ComponentRenderer {
  constructor(
    component: Component,
    assets: Assets,
    propOverridesMap: Map<string, ComponentProps>
  ) {
    this.project = component.project;
    this.component = component;
    this.assets = assets;
    this.propOverridesMap = propOverridesMap;
  }

  readonly project: Project;
  readonly component: Component;
  readonly assets: Assets;
  readonly propOverridesMap = new Map<string, ComponentProps>();

  getPropOverridesForNode(node: Node) {
    if (node.id === this.component.rootNode.id) {
      return this.propOverridesMap.get("") ?? {};
    }
    const refID = this.component.refIDs.get(node.id);
    if (!refID) {
      return {};
    }
    return this.propOverridesMap.get(refID) ?? {};
  }

  render(instancePath: string[]): JSX.Element {
    return this.renderNode(this.component.rootNode, instancePath, true);
  }

  renderNode(node: Node, instancePath: string[], isRoot: boolean): JSX.Element {
    const isInstanceRoot = isRoot && instancePath.length;

    const idPath = isInstanceRoot ? instancePath : [...instancePath, node.id];
    const style = this.project.getStyle(idPath);

    const tagName = style.tagName ?? "div";

    const classNames: string[] = [];
    for (let i = 0; i < idPath.length; ++i) {
      classNames.push("uimix-" + idPath.slice(i).join("-"));
    }
    if (isInstanceRoot) {
      classNames.push("uimix-" + node.id);
    }

    const overrideProps = this.getPropOverridesForNode(node);
    const commonProps: ComponentProps = {
      className: classNames.join(" "),
      ...overrideProps,
    };
    delete commonProps.overrides;

    const hasChildrenOverride = overrideProps.children !== undefined;

    if (node.type === "text") {
      return React.createElement(
        tagName,
        { ...commonProps },
        ...(hasChildrenOverride ? [] : [style.textContent])
      );
    }
    if (node.type === "image") {
      const src = style.imageHash;
      const imageURL = src ? this.assets.imageURLs.get(src) : undefined;
      return React.createElement("img", { ...commonProps, src: imageURL });
    }
    if (node.type === "instance") {
      const mainComponentID = style.mainComponentID;
      const componentRenderer =
        mainComponentID &&
        this.component.project.components.get(mainComponentID);
      if (!componentRenderer) {
        console.error(
          `Component ${mainComponentID} not found for instance ${node.id}`
        );
        return React.createElement("div");
      }

      const { overrides, ...rootProps } = overrideProps;
      const innerOverridesMap = new Map(Object.entries(overrides ?? {}));
      innerOverridesMap.set("", rootProps);

      return new ComponentRenderer(
        componentRenderer,
        this.assets,
        innerOverridesMap
      ).render([...instancePath, node.id]);
    }
    if (node.type === "foreign") {
      const foreignComponentID = style.foreignComponentID;
      if (!foreignComponentID) {
        console.error(`Foreign component ID not found for node ${node.id}`);
        return React.createElement("div");
      }

      const foreignComponent = this.assets.foreignModules.get(
        foreignComponentID.path
      )?.[foreignComponentID.name];
      if (!foreignComponent) {
        console.error(
          `Foreign component ${foreignComponentID.path}#${foreignComponentID.name} not found for node ${node.id}`
        );
        return React.createElement("div");
      }

      return React.createElement(
        tagName,
        { ...commonProps },
        React.createElement(foreignComponent, { ...foreignComponentID.props })
      );
    }

    return React.createElement(
      tagName,
      { ...commonProps },
      ...(hasChildrenOverride
        ? []
        : node.children.map((child) =>
            this.renderNode(child, instancePath, false)
          ))
    );
  }
}

function createReactComponent(
  component: Component,
  assets: Assets
): React.FC<ComponentProps> {
  const Component = React.forwardRef((props, ref) => {
    const { overrides, ...rootOverride } = { ...props, ref };
    const propOverridesMap = new Map<string, ComponentProps>();
    for (const [refID, override] of Object.entries(overrides ?? {})) {
      propOverridesMap.set(refID, override);
    }
    propOverridesMap.set("", rootOverride);

    return new ComponentRenderer(component, assets, propOverridesMap).render(
      []
    );
  }) as React.FC<ComponentProps>;
  Component.displayName = component.name;
  return Component;
}
