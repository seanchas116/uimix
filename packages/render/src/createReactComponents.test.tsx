import React from "react";
import { describe, expect, it } from "vitest";
import reactRenderer from "react-test-renderer";
import { createReactComponents } from "./createReactComponents";
// @ts-ignore
import projectJSON from "./__fixtures__/uimix/data.json";

describe(createReactComponents.name, () => {
  it("generates components", async () => {
    const components = createReactComponents(
      projectJSON,
      {
        "YmY-tuteU03QY-k_2ARy1uacNmXDNmOX-FFAlj2qCIo=": (
          await import(
            // @ts-ignore
            "./__fixtures__/uimix/images/YmY-tuteU03QY-k_2ARy1uacNmXDNmOX-FFAlj2qCIo=.png"
          )
        ).default,
      },
      {
        "src/stories/Button.tsx": await import(
          "../../sandbox/src/stories/Button"
        ),
        "src/stories/Header.tsx": await import(
          "../../sandbox/src/stories/Header"
        ),
      }
    );

    console.log(components);

    const Button = components.get("Button")!;
    const Stack = components.get("Stack")!;
    const Page = components.get("Page")!;

    expect(
      reactRenderer
        .create(
          <Button
            overrides={{
              text: {
                children: "Hello, world!",
              },
            }}
          />
        )
        .toJSON()
    ).toMatchSnapshot();
    expect(reactRenderer.create(<Stack />).toJSON()).toMatchSnapshot();
    expect(
      reactRenderer
        .create(
          <Page
            overrides={{
              stack: {
                overrides: {
                  text: {
                    children: "Nested oveerride!",
                  },
                },
              },
            }}
          />
        )
        .toJSON()
    ).toMatchSnapshot();
  });
});
