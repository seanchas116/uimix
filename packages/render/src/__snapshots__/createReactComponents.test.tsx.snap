// Vitest Snapshot v1

exports[`createReactComponents > generates components 1`] = `
<div
  className="uimix-1Bvi2z3UPfi6W3L8nDt5VA"
>
  <div
    className="uimix-1Bvi2z3UGTM64JtRT7Qc1S"
  >
    Hello, world!
  </div>
</div>
`;

exports[`createReactComponents > generates components 2`] = `
<div
  className="uimix-1Bvi2z3TfkSpYXW29daTN8"
>
  <div
    className="uimix-1Bvi2z3TpKXtTJLTRPTEd8"
  />
  <div
    className="uimix-1Bvi2z3To8ddXo6R6ZftVD"
  />
  <div
    className="uimix-1Bvi2z3UB1xUVg4Xz9WTpJ"
  />
  <div
    className="uimix-1Bvi2z3U3SiK6r7gf9vh7N"
  >
    Hello, world!
  </div>
</div>
`;

exports[`createReactComponents > generates components 3`] = `
<div
  className="uimix-1Bvi319YVxZELz4tEskN1S"
>
  <div
    className="uimix-1Bvi31VKu8gfy6XzJ8xrHA"
  >
    <header>
      <div
        className="wrapper"
      >
        <div>
          <svg
            height="32"
            viewBox="0 0 32 32"
            width="32"
            xmlns="http://www.w3.org/2000/svg"
          >
            <g
              fill="none"
              fillRule="evenodd"
            >
              <path
                d="M10 0h12a10 10 0 0110 10v12a10 10 0 01-10 10H10A10 10 0 010 22V10A10 10 0 0110 0z"
                fill="#FFF"
              />
              <path
                d="M5.3 10.6l10.4 6v11.1l-10.4-6v-11zm11.4-6.2l9.7 5.5-9.7 5.6V4.4z"
                fill="#555AB9"
              />
              <path
                d="M27.2 10.6v11.2l-10.5 6V16.5l10.5-6zM15.7 4.4v11L6 10l9.7-5.5z"
                fill="#91BAF8"
              />
            </g>
          </svg>
          <h1>
            Acme
          </h1>
        </div>
        <div>
          <button
            className="storybook-button storybook-button--small storybook-button--secondary"
            style={
              {
                "backgroundColor": undefined,
              }
            }
            type="button"
          >
            Log in
          </button>
          <button
            className="storybook-button storybook-button--small storybook-button--primary"
            style={
              {
                "backgroundColor": undefined,
              }
            }
            type="button"
          >
            Sign up
          </button>
        </div>
      </div>
    </header>
  </div>
  <div
    className="uimix-1Bvi365mtWZmXKDPeJTB4s"
  >
    <div
      className="uimix-1Bvi32d9RMi3ES5kk2UvuM"
    >
      <button
        className="storybook-button storybook-button--large storybook-button--primary"
        style={
          {
            "backgroundColor": "red",
          }
        }
        type="button"
      >
        Button1
      </button>
    </div>
    <div
      className="uimix-1Bvi2z3UrU9VkawwAteXs8"
    >
      <div
        className="uimix-1Bvi2z3UrU9VkawwAteXs8-1Bvi2z3UGTM64JtRT7Qc1S"
      >
        Override Name
      </div>
    </div>
    <img
      className="uimix-1Bvi392qBdMy93LGLcVymb"
      src="/src/__fixtures__/uimix/images/YmY-tuteU03QY-k_2ARy1uacNmXDNmOX-FFAlj2qCIo=.png"
    />
    <div
      className="uimix-1Bvi39mct9UNYZ9eTCCRdV"
    >
      <div
        className="uimix-1Bvi39mct9UNYZ9eTCCRdV-1Bvi2z3TpKXtTJLTRPTEd8"
      />
      <div
        className="uimix-1Bvi39mct9UNYZ9eTCCRdV-1Bvi2z3To8ddXo6R6ZftVD"
      />
      <div
        className="uimix-1Bvi39mct9UNYZ9eTCCRdV-1Bvi2z3UB1xUVg4Xz9WTpJ"
      />
      <div
        className="uimix-1Bvi39mct9UNYZ9eTCCRdV-1Bvi2z3U3SiK6r7gf9vh7N"
      >
        Nested oveerride!
      </div>
    </div>
  </div>
</div>
`;
